### ISSUE TITLE
A combination of "pipe", "chain", "map" and "reduce" keeps the "accumulator" between function calls

### ISSUE THREAD

**Issue by dragos-rosca** [ramda/ramda#3498]
Hello,

I tried to extract some fields from a list with some suboptimal code that revealed a bug, I think.

<details>
<summary>
[Click Me] Given the following data:
</summary>

```javascript
const frs = [
  {
    id: 'c86a051a-1fec-47b6-89a8-1cee0b834ce3',
    code: 'tenant_admin',
    name: 'Tenant Admin',
    description: null,
    permissions: [
      {
        id: '55fffa07-4d59-46a7-a81a-542fcb09a904',
        functionalRoleId: 'c86a051a-1fec-47b6-89a8-1cee0b834ce3',
        permissionId: 'f9153a09-2a2b-4d7a-bea0-af1c4f1313e0',
        permission: {
          id: 'f9153a09-2a2b-4d7a-bea0-af1c4f1313e0',
          code: 'CRUD.Parameter',
          name: 'Change parameter',
          description: 'Create, read, update or delete a Parameter.',
          isEnvDependent: true
        },
        environmentGroups: [
          {
            id: '2cb10595-fc43-4d39-849a-1532cd2367db',
            environmentGroupId: 'f5961c69-8711-412c-aea1-05194bb54b04',
            functionalRolePermissionId: '55fffa07-4d59-46a7-a81a-542fcb09a904',
            environmentGroup: {
              id: 'f5961c69-8711-412c-aea1-05194bb54b04',
              name: 'UAT',
              environments: [
                {
                  id: '99ef6b89-3aca-4249-9dba-b3296e9170c8',
                  name: 'UAT1',
                  description: 'TEST',
                  code: 'UAT1',
                  portalId: null,
                  stateId: 'add8a39a-0644-4da9-ae5f-044e6e4da70d',
                  environmentGroupId: 'f5961c69-8711-412c-aea1-05194bb54b04'
                }
              ]
            }
          },
          {
            id: '66bf5649-0e48-48d8-bc5e-3a867b24197a',
            environmentGroupId: '923e02c6-e416-4cb5-a8d5-77f8d0e648d8',
            functionalRolePermissionId: '55fffa07-4d59-46a7-a81a-542fcb09a904',
            environmentGroup: {
              id: '923e02c6-e416-4cb5-a8d5-77f8d0e648d8',
              name: 'PROD',
              environments: [
                {
                  id: 'e14b721e-5ee8-4313-b630-1f5fca6497a9',
                  name: 'PROD1',
                  description: 'PROD1',
                  code: 'PROD1',
                  portalId: null,
                  stateId: 'add8a39a-0644-4da9-ae5f-044e6e4da70d',
                  environmentGroupId: '923e02c6-e416-4cb5-a8d5-77f8d0e648d8'
                }
              ]
            }
          }
        ]
      }
    ]
  },
  {
    id: 'afc6dc05-1921-42b4-97df-2f1cea8ff32d',
    code: 'tenant_user',
    name: 'Tenant User',
    description: 'This is a default role that the identity gives to new comers!',
    permissions: [
      {
        id: 'd905e07b-b7ea-4e1f-bdfd-7b9d360622d8',
        functionalRoleId: 'afc6dc05-1921-42b4-97df-2f1cea8ff32d',
        permissionId: 'f9153a09-2a2b-4d7a-bea0-af1c4f1313e0',
        permission: {
          id: 'f9153a09-2a2b-4d7a-bea0-af1c4f1313e0',
          code: 'CRUD.Parameter',
          name: 'Change parameter',
          description: 'Create, read, update or delete a Parameter.',
          isEnvDependent: true
        },
        environmentGroups: [
          {
            id: 'bc1adafa-f4c2-468a-8e36-64b9f1a039f8',
            environmentGroupId: '99ffcebf-34b8-4cfe-8cba-d523166160a9',
            functionalRolePermissionId: 'd905e07b-b7ea-4e1f-bdfd-7b9d360622d8',
            environmentGroup: {
              id: '99ffcebf-34b8-4cfe-8cba-d523166160a9',
              name: 'DEV',
              environments: [
                {
                  id: 'ecbfec2a-f1a0-4d9a-bcd2-338f412ec338',
                  name: 'DEV1',
                  description: 'test',
                  code: 'DEV1',
                  portalId: null,
                  stateId: 'add8a39a-0644-4da9-ae5f-044e6e4da70d',
                  environmentGroupId: '99ffcebf-34b8-4cfe-8cba-d523166160a9'
                },
                {
                  id: '0ab85ab3-1b96-49e5-b566-e2477a962b33',
                  name: 'DEV2',
                  description: 'DEV2',
                  code: 'DEV2',
                  portalId: null,
                  stateId: 'add8a39a-0644-4da9-ae5f-044e6e4da70d',
                  environmentGroupId: '99ffcebf-34b8-4cfe-8cba-d523166160a9'
                }
              ]
            }
          }
        ]
      },
      {
        id: 'a9a60e82-d08c-42fd-9f11-8f32d779af95',
        functionalRoleId: 'afc6dc05-1921-42b4-97df-2f1cea8ff32d',
        permissionId: 'a2600310-883c-4022-9cb4-459b0cc345ad',
        permission: {
          id: 'a2600310-883c-4022-9cb4-459b0cc345ad',
          code: 'CRUD.FunctionalRole',
          name: 'Change Functional Role',
          description: 'Create, read, update or delete a Functional role.',
          isEnvDependent: false
        },
        environmentGroups: []
      }
    ]
  }
]
```
</details>

TLDR: We have some roles that have a list of permissions, each permission has a list of environment groups and each environment group has a list of environments. We wanted to extract a list of (unique) permissions with a merged list of environments (I know that the environments need to be unique too but that would have hidden the bug).

I used the following function to extract what I needed
```javascript
const extractPermissions = pipe(
  chain(prop('permissions')),
  map(permission => ({
    code: path(['permission', 'code'], permission),
    environments: pipe(chain(prop('environments')), pluck('id'))(chain(prop('environmentGroup'), permission.environmentGroups))
  })),
  reduce((acc, curr) => {
    const existing = acc.find(item => item.code === curr.code)
    if (existing) {
      existing.environments = concat(existing.environments, curr.environments)
    } else {
      acc.push(curr)
    }
    return acc
  }, [])
)
```

And to test my code I tried a simple `for` statement to measure performance. When I saw I broke the REPL on the Ramda site at only 100k iterations I thought it was suspicious.
After some investigations I found out that somehow the accumulator of the `reduce` is being reused somehow giving the following output:
```javascript
console.log(extractPermissions(frs))
// Output: [{"code":"CRUD.Parameter","environments":["99ef6b89-3aca-4249-9dba-b3296e9170c8","e14b721e-5ee8-4313-b630-1f5fca6497a9","ecbfec2a-f1a0-4d9a-bcd2-338f412ec338","0ab85ab3-1b96-49e5-b566-e2477a962b33"]},{"code":"CRUD.FunctionalRole","environments":[]}]
console.log(extractPermissions(frs))
// Output: [{"code":"CRUD.Parameter","environments":["99ef6b89-3aca-4249-9dba-b3296e9170c8","e14b721e-5ee8-4313-b630-1f5fca6497a9","ecbfec2a-f1a0-4d9a-bcd2-338f412ec338","0ab85ab3-1b96-49e5-b566-e2477a962b33","99ef6b89-3aca-4249-9dba-b3296e9170c8","e14b721e-5ee8-4313-b630-1f5fca6497a9","ecbfec2a-f1a0-4d9a-bcd2-338f412ec338","0ab85ab3-1b96-49e5-b566-e2477a962b33"]},{"code":"CRUD.FunctionalRole","environments":[]}]
```

And for each new call of the function the line `existing.environments = concat(existing.environments, curr.environments)` keeps adding new data to the field.

I am not interested in optimizing the `extractPermissions` function, I already fixed it and used a completely different approach. 

Please help me understand how is the `acc` kept between calls. How making a function with Ramda is not pure?

I think this is a bug, but it is possible I made some mistake that I can't see.

**Comment by bunglegrind** [comment]
From the reduce documentation:

> Be cautious of mutating and returning the accumulator. If you reuse it across invocations, it will continue to accumulate onto the same value. The general recommendation is to always return a new value. If you can't do so for performance reasons, then be sure to reinitialize the accumulator on each invocation.

I can see:

```
acc.push(curr)
return acc
```

**Comment by Harris-Miller** [comment]
> so maybe it works as it's suppose to?

Your problem is one of javascript, and not of ramda specifically. You don't even need the `pipe` to produce the undesired behavior:
```typescript
const x = [1,2,3]
const fn = reduce((acc, i) => (acc.a += i, acc), { a: 0 })

console.log(fn(x))
//=> {"a":6}
console.log(fn(x))
//=> {"a":12}
console.log(fn(x))
//=> {"a":18}
```

The problem here is closures. `reduce` is curried so your accumulator object is part of the closure scope for `fn`. Each call of `fn` references that object. It is not re-declaring it each time

You'd get the same thing by deconstructing `reduce` into a manually curried arrow function and a native array#reduce

```typescript
const x = [1,2,3]
const reduce2 = (fn, acc) => l => l.reduce(fn, acc);
const fn = reduce2((acc, i) => (acc.a += i, acc), { a: 0 })

console.log(fn(x))
//=> {"a":6}
console.log(fn(x))
//=> {"a":12}
console.log(fn(x))
//=> {"a":18}
```

Wrapping either `R.reduce` or a native Array#reduce in an arrow function would perform a new evaluation, creating a new object literal for each run

```typescript
const x = [1,2,3]
const fn = l => reduce((acc, i) => (acc.a += i, acc), { a: 0 }, l)

console.log(fn(x))
//=> {"a":6}
console.log(fn(x))
//=> {"a":6}
console.log(fn(x))
//=> {"a":6}
```
or
```typescript
const x = [1,2,3]
const fn = l => l.reduce((acc, i) => (acc.a += i, acc), { a: 0 })

console.log(fn(x))
//=> {"a":6}
console.log(fn(x))
//=> {"a":6}
console.log(fn(x))
//=> {"a":6}
```

**Comment by dragos-rosca** [comment]
Thank you all for your responses.
I see that the problem is the closure and more importantly the problem is that I break the "immutability" and "purity" rule of FP.

In FP you should never alter (mutate) the parameters.

I will close this issue but I would like to make a suggestion: 
• Can we put something like `Object.freeze` behind the scene between iteration calls for the accumulator?
It will at least throw an error like:
![image](https://github.com/user-attachments/assets/2b9c9b43-2477-4542-b699-5adc1c2fa7b2)

**Comment by Harris-Miller** [comment]
> Be cautious of mutating and returning the accumulator. If you reuse it across invocations, it will continue to accumulate onto the same value. The general recommendation is to always return a new value. If you can't do so for performance reasons, then be sure to reinitialize the accumulator on each invocation.

Sorry mean to address this too... As you said, this is referring to the `acc.push(curr)` as you pointed out. So if you didn't want to solve the problem of closures here with an arrow wrapper function, you'd want to no mutate the array direct (you can still mutate items in the arrow though

```typescript
const extractPermissions = pipe(
  chain(prop('permissions')),
  map(permission => ({
    code: path(['permission', 'code'], permission),
    environments: pipe(chain(prop('environments')), pluck('id'))(chain(prop('environmentGroup'), permission.environmentGroups))
  })),
  reduce((acc, curr) => {
    const existing = acc.find(item => item.code === curr.code)
    if (existing) {
      existing.environments = concat(existing.environments, curr.environments)
      return acc;
    }

    return [...acc, curr];
  }, [])
)
```
See how this will use the same instance of `[]` for all iterations, but you also return a new one when adding a new item? That means each invocation will start with an emptyArray
